package/libskylark/src/libskylark.c

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright (C) 2017 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <curl/curl.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "libskylark.h"
+
+/**
+ * Functions and utilities related to interfacing with Swift Skylark service.
+ *
+ * This library defines a suite of shared functionality required for all clients
+ * to Skylark, containing some utility functions around reading, managing, and
+ * displaying client configurations, as well as libcurl-based requests for SBP
+ * publish and subscribe of SBP data over HTTP.
+ */
+
+/**
+ *  Utilities
+ */
+
+// TODO (mookerji): Repeated here as is everywhere else! This needs to be moved
+// into a piksi buildroot library.
+static int file_read_string(const char *filename, char *str, size_t str_size)
+{
+  FILE *fp = fopen(filename, "r");
+  if (fp == NULL) {
+    log_error("error opening %s\n", filename);
+    return -1;
+  }
+  bool success = (fgets(str, str_size, fp) != NULL);
+  fclose(fp);
+  if (!success) {
+    log_error("error reading %s\n", filename);
+    return -1;
+  }
+  return 0;
+}
+
+/** Serialize client return code to a human readable message.
+ *
+ * \param code  Client return code
+ * \return  Error message
+ */
+const char *client_strerror(RC code)
+{
+  switch (code) {
+    case NO_ERROR:
+      return "No error";
+    case E_NOT_IMPLEMENTED:
+      return "Not implemented";
+    case E_GENERIC_ERROR:
+      return "Generic error";
+    case E_NULL_VALUE_ERROR:
+      return "Required value was NULL";
+    case E_CONF_READ_ERROR:
+      return "Client: Error reading client configuration parameters";
+    case E_INITIALIZATION_ERROR:
+      return "Client: Initialization error";
+    case E_NETWORK_UNAVAILABLE:
+      return "Client: The network is unavailable";
+    case E_CONNECTION_LOST:
+      return "Client: The network connection was lost";
+    case E_RECONNECTION_FAILED:
+      return "Client: Network reconnection failed";
+    case E_SUB_CONNECTION_ERROR:
+      return "Client: Subscribe connection failed";
+    case E_SUB_WRITE_ERROR:
```

Comment:
Some of these may seem unnecessary, but I believe they can be added with the read/write handling in the callbacks.